xonsh/tools.py

```diff
@@ -1634,6 +1647,38 @@ def lazyget(self, key, default=None):
         """A lazy value getter."""
         return self._cmds_cache.get(key, default)
 
+    def locate_binary(self, name):
+        """Locates an executable on the file system."""
+
+        if ON_WINDOWS:
+            # Windows users expect to be able to execute files in the same
+            # directory without `./`
+            cwd = _get_cwd()
+
+            local_bins = [full_name for full_name in self.get_possible_names(name)
+                          if os.path.isfile(full_name)]
+            if local_bins:
+                return os.path.abspath(os.path.relpath(local_bins[0], cwd))
+
+        possibilities = self.get_possible_names(name)
+        matches = possibilities & self.all_commands.keys()
+
+        if matches:
+            return self.all_commands[matches.pop()][0]
+        elif os.path.isfile(name) and name != os.path.basename(name):
+            return name
+
+    def lazy_locate_binary(self, name):
+        """Locates an executable in the cache."""
+
+        possibilities = self.get_possible_names(name)
+        matches = possibilities & self._cmds_cache.keys()
```

Comment:
This is a pretty inefficient way to check if a single name is in the cache.  Assuming possibilities returns a list or names, a faster implementation would be:

```python
match = None
for n in self.get_possible_names(name):
    if n in self._cmds_cache:
        match = n
        break
```